### PR TITLE
security: Stabilize AdvancedTlsX509KeyManager.

### PR DESCRIPTION
**Pull request by erm-g**
This PR is a part of 'Stabilize Advanced TLS' effort.
Clean up, improve javadoc, de-experimentalize of AdvancedTlsX509KeyManager, add a unit test (e2e already exists).

### CODE REVIEW

File: util/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java
Comment (matthewstevenson88) [inline util/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java:51]: nit: Please use `Duration` here.

Comment (matthewstevenson88) [inline util/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java:51]: Should this be marked `private`?

**Review by matthewstevenson88 — COMMENTED** [review]
Thanks for the PR @erm-g! A couple nits, and two other points:

1. This PR only de-experimentalizes the key manager. Should we amend the PR title accordingly?
2. Please address the failing tests.

File: util/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java
Comment (erm-g) [reply util/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java]: Since JUL is used, most of the times it's configured to print method as well, i.e.
`io.grpc.util.AdvancedTlsX509KeyManager updateIdentityCredentialsFromFile
INFO: Provided refresh period...`
IMHO it's already specific enough for a user to understand the issue. WDYT?

File: util/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java
Comment (erm-g) [reply util/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java]: I'd prefer to keep it public since I see few other constants are public as well, plus it's a part of implementation that we don't hide from the users.

Comment (erm-g) [reply util/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java]: They use 'period' across the class here, even in public API like `updateIdentityCredentialsFromFile(File keyFile, File certFile, long period, TimeUnit unit...` so IMHO it's better fits existing style

Comment (matthewstevenson88) [reply util/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java]: I'll defer to Eric here, but the general Google Java style is to not use primitives to represent time.

Comment (matthewstevenson88) [reply util/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java]: Can you point to these other examples?

From my point of view, we should not mark a field public unless there is a clear reason why a user needs to depend on this value. In this case, it's not clear to me what that reason is.

File: util/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java
Comment (matthewstevenson88) [reply util/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java]: SGTM.

**Review by matthewstevenson88 — COMMENTED** [review]
Please address the failing tests. :)

File: util/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java
Comment (erm-g) [reply util/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java]: The following 2 are public and I couldn't find any usages outside of tests. Maybe there is a reason why a user needs to depend on them but I'm not sure

https://github.com/grpc/grpc-java/blob/2bc4306940e7a2dea529be8a6ce536cab0638dc0/core/src/main/java/io/grpc/internal/GrpcUtil.java#L189

https://github.com/grpc/grpc-java/blob/2bc4306940e7a2dea529be8a6ce536cab0638dc0/alts/src/main/java/io/grpc/alts/internal/AltsTsiHandshaker.java#L39

**Comment by erm-g** [comment]
> Please address the failing tests. :)

Fixed (few styling things)

File: util/src/test/java/io/grpc/util/AdvancedTlsX509KeyManagerTest.java
Comment (matthewstevenson88) [inline util/src/test/java/io/grpc/util/AdvancedTlsX509KeyManagerTest.java:52]: micro-nit: Please mark these constants as `private`, as there is no need for them to be public.

File: util/src/test/java/io/grpc/util/AdvancedTlsX509KeyManagerTest.java
Comment (matthewstevenson88) [inline util/src/test/java/io/grpc/util/AdvancedTlsX509KeyManagerTest.java:108]: We seem to be testing several distinct behaviors in the same unit test - can we break these up into separate smaller-scoped tests so that each unit test is testing an independent behavior?

Similarly above in the `credentialSetting` test.

File: util/src/test/java/io/grpc/util/AdvancedTlsX509KeyManagerTest.java
Comment (matthewstevenson88) [inline util/src/test/java/io/grpc/util/AdvancedTlsX509KeyManagerTest.java:72]: optional: Consider simply `throws Exception` here, rather than throwing all 4. Since this is a unit test, we don't need to be super careful to throw narrowly-scoped exceptions.

File: util/src/test/java/io/grpc/util/AdvancedTlsX509KeyManagerTest.java
Comment (matthewstevenson88) [inline util/src/test/java/io/grpc/util/AdvancedTlsX509KeyManagerTest.java:1]: Disclaimer: I'm very much in favor of more testing: )

How do you view the roles of this test file compared to `AdvancedTlsTest.java`? Is the goal of this test file to be the unit tests, whereas `AdvancedTlsTest.java` consists of the e2e-style tests? If yes, should we be comprehensive in this file and ensure that all methods in `AdvancedTlsX509KeyManager` are adequately unit-tested?

File: util/src/test/java/io/grpc/util/AdvancedTlsX509KeyManagerTest.java
Comment (matthewstevenson88) [inline util/src/test/java/io/grpc/util/AdvancedTlsX509KeyManagerTest.java:90]: nit: The Google Java style guide recommends Java test names follow the pattern `<method-under-test>_<expected-behavior>`, so e.g. the first stanza in this test could be grouped into a unit test called `updateIdentityCredentials_success`. I'm not sure how strictly gRPC-Java follows the Google Java style guide, but wanted to call this out in case.

https://google.github.io/styleguide/javaguide.html#s5.2.3-method-names

File: util/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java
Comment (matthewstevenson88) [reply util/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java]: I think there are is a difference between those examples and this one, because those libraries are very much internal to gRPC, whereas this is a public API. Further, in the `AltsTsiHandshaker.java` case it really should be package-private and annotated with the "VisibleForTesting" annotation.

For a public API such as this one, I think we should not expose a constant like this one to the user unless we expect them to really need to have access to it. If there is such a reason in this case, please let me know. Otherwise, I do suggest we mark this private.

File: util/src/test/java/io/grpc/util/AdvancedTlsX509KeyManagerTest.java
Comment (erm-g) [reply util/src/test/java/io/grpc/util/AdvancedTlsX509KeyManagerTest.java:108]: In this test we're just checking that we correctly put `checkNotNull` checks across the public api. So I grouped that under single unit test since I saw a similar pattern at MatcherTest - 
https://github.com/grpc/grpc-java/blob/2bc4306940e7a2dea529be8a6ce536cab0638dc0/xds/src/test/java/io/grpc/xds/internal/MatcherTest.java#L64

`credentialSetting` is different - we check a sequence of changing (serverCert->clientCert->serverCert) so splitting it is possible, but will require a lot of boilerplate code (smth like wordy BeforeMethod before each test)

File: util/src/test/java/io/grpc/util/AdvancedTlsX509KeyManagerTest.java
Comment (erm-g) [reply util/src/test/java/io/grpc/util/AdvancedTlsX509KeyManagerTest.java:1]: Yes, `AdvancedTlsTest.java` is testing the approach from e2e perspective, involves both Key and Trust manager, and, IMHO is more useful for the users. 
While looking at it I realized that some of the minor cases (like NPEs) are not covered, and LOC like 
https://github.com/grpc/grpc-java/blob/2bc4306940e7a2dea529be8a6ce536cab0638dc0/netty/src/test/java/io/grpc/netty/AdvancedTlsTest.java#L341

are confusing - no matter what timeout you specify, the initial call for loading creds is synchronous.
So I ended up with some extra unit test coverage of API itself (plus updating javadoc), plus checking that our minimum period is really enforced.
Unit testing all methods and cases will be a)very complicated because of bunch of private/final things (even mockito struggles a lot with it, for example, look at how to emulate a FS reading error of a keyFile) b)not fitting the overall style of the repo (I couldn't see such super detailed tests)

File: util/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java
Comment (erm-g) [reply util/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java]: fixed based on offline talk

File: util/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java
Comment (ejona86) [inline util/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java:41]: Why was this restriction added? I know we talked about it some, but you can totally call some of these multiple times, and that's okay.

File: util/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java
Comment (ejona86) [inline util/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java:51]: Include "MINUTES" in the name?

File: util/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java
Comment (ejona86) [inline util/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java:136]: We never use INFO, because essentially nobody ever wants to see them. Either it is a problem or it isn't, and if it is a problem, throw an exception. We could make this FINE, but I don't think we should have it INFO which is logged by default. Logging at INFO would be the same as saying it is against the API.

File: util/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java
Comment (ejona86) [inline util/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java:142]: What's the goal with this checkNotNull being added?

File: util/src/test/java/io/grpc/util/AdvancedTlsX509KeyManagerTest.java
Comment (ejona86) [reply util/src/test/java/io/grpc/util/AdvancedTlsX509KeyManagerTest.java:90]: We do use the underscore notation when it is helpful. But it isn't a strict requirement. Ever from the style guide you linked it isn't a strict requirement; it very explicitly says "There is no One Correct Way to name test methods." And our code is probably older than those parts of the style guide.

File: util/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java
Comment (erm-g) [reply util/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java:142]: I saw a 'checkNotNull' approach in this class for one public API - 
https://github.com/grpc/grpc-java/blob/8aaace12eb8697a2dd841d2261b0051a1414f0c8/util/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java#L111

and just added it for other public API methods. In this particular case without it we'd get a raw NPE at 
https://github.com/grpc/grpc-java/blob/8aaace12eb8697a2dd841d2261b0051a1414f0c8/util/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java#L135
if the user passes null for 'executor' param.

File: util/src/test/java/io/grpc/util/AdvancedTlsX509KeyManagerTest.java
Comment (ejona86) [reply util/src/test/java/io/grpc/util/AdvancedTlsX509KeyManagerTest.java:1]: > no matter what timeout you specify, the initial call for loading creds is synchronous.

Oh, we should use `FakeClock` for that. Then we can advance time manually within the test and only have the one test thread.

File: util/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java
Comment (erm-g) [reply util/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java:136]: Oh, I see that it's FINE in a lot of similar places. Thank you, changed.

File: util/src/test/java/io/grpc/util/AdvancedTlsX509KeyManagerTest.java
Comment (ejona86) [reply util/src/test/java/io/grpc/util/AdvancedTlsX509KeyManagerTest.java:72]: For tests, it is much better to `throws Exception`. The specific exception fluctuate and doesn't provide value, as you are asserting they never happen.

File: util/src/test/java/io/grpc/util/AdvancedTlsX509KeyManagerTest.java
Comment (ejona86) [reply util/src/test/java/io/grpc/util/AdvancedTlsX509KeyManagerTest.java:52]: +1 to private. AdvancedTlsTest was written by a non-Java programmer. Some of the style was noticed/fixed, but it's very believable things snuck through.

File: util/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java
Comment (ejona86) [reply util/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java]: gRPC code pretty much never uses Duration because we support Android versions that don't have it, and we don't require API-level desugaring. If a specific package doesn't support Android, then it could potentially use it, but it is easier to not use it anywhere. (Streams are probably the one case that are used some when able and not elsewhere. Not to prefer streams, but the author found it valuable.)

File: util/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java
Comment (erm-g) [reply util/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java]: Changed to MINIMUM_REFRESH_PERIOD_IN_MINUTES

File: util/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java
Comment (erm-g) [reply util/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java:41]: Agreed, it's perfectly fine to call updateIdentityCredentials and updateIdentityCredentialsFromFile(File, File) (unless a user does it from multiple threads) multiple times.
I think that calling those 2 after updateIdentityCredentialsFromFile (File, File, long, TimeUnit, ScheduledExecutorService) is something we want to discourage - WDYT?

Comment (ejona86) [reply util/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java:41]: As long as you call `close()` on the return value of `updateIdentityCredentialsFromFile(File, File, long, TimeUnit, ScheduledExecutorService)`, all the methods are reasonable again.

We can add to that one method "you must close() before calling other update methods"

File: util/src/test/java/io/grpc/util/AdvancedTlsX509KeyManagerTest.java
Comment (erm-g) [reply util/src/test/java/io/grpc/util/AdvancedTlsX509KeyManagerTest.java]: Fixed here and cleqned up exceptions for AdvancedTlsTest

File: util/src/test/java/io/grpc/util/AdvancedTlsX509KeyManagerTest.java
Comment (erm-g) [reply util/src/test/java/io/grpc/util/AdvancedTlsX509KeyManagerTest.java]: Fixed here and for AdvancedTlsTest as well

File: util/src/test/java/io/grpc/util/AdvancedTlsX509KeyManagerTest.java
Comment (erm-g) [reply util/src/test/java/io/grpc/util/AdvancedTlsX509KeyManagerTest.java:1]: > Oh, we should use `FakeClock` for that.

Agreed. Right now there is no test covering time based reloading from file - the current one just relies on the fact that the `updateIdentityCredentialsFromFile(File keyFile, File certFile, long period, TimeUnit unit, ScheduledExecutorService executor)` calls `readAndUpdate` before scheduling the periodic updates - 
https://github.com/grpc/grpc-java/blob/8aaace12eb8697a2dd841d2261b0051a1414f0c8/util/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java#L129

I can cover it in a separate PR and do it in probably e2e AdvancedTlsTest.

File: util/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java
Comment (erm-g) [reply util/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java]: Makes sense, agreed. For example, if a user wants to

1. load initial creds from somewhere
2. keep updating creds from file
3. reload the creds urgently
4. continue updating creds from file
then it can be achieved by the following sequence

1. `updateIdentityCredentials`
2. `updateIdentityCredentialsFromFile(File, File, long, TimeUnit, ScheduledExecutorService)`
3. `close`
4. `updateIdentityCredentials`
5. `updateIdentityCredentialsFromFile(File, File, long, TimeUnit, ScheduledExecutorService)`

Updated the wording

File: util/src/test/java/io/grpc/util/AdvancedTlsX509KeyManagerTest.java
Comment (erm-g) [reply util/src/test/java/io/grpc/util/AdvancedTlsX509KeyManagerTest.java:87]: @ejona86 just want to double check - are you ok with the current names, or do you want me to change it to underscore pattern?

File: util/src/test/java/io/grpc/util/AdvancedTlsX509KeyManagerTest.java
Comment (ejona86) [reply util/src/test/java/io/grpc/util/AdvancedTlsX509KeyManagerTest.java:1]: I'm fine with you adding the test in a separate PR, but let's swap to FakeClock now, to never use additional threads. (`new FakeClock().getScheduledExecutorService()`)

**Review by ejona86 — APPROVED** [review]
Let's swap to FakeClock, but otherwise looks good.

**Comment by ejona86** [comment]
@matthewstevenson88, do you want us to wait for your approval before this goes in?

**Review by matthewstevenson88 — APPROVED** [review]

**Comment by matthewstevenson88** [comment]
Thanks @ejona86. LGTM, and ok to merge once the FakeClock change is done.

**Comment by ejona86** [comment]
API review meeting notes:

> an object that caller should close when the file refreshes are not needed

s/not/no longer/

We noticed we had looked at this API before, and had talked about changing the argument order. https://github.com/grpc/grpc-java/issues/8024#issuecomment-1220003981 . Although I did today notice that KeyStore uses the argument order seen here. https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/security/KeyStore.html#setKeyEntry(java.lang.String,byte%5B%5D,java.security.cert.Certificate%5B%5D)

We probably would let y'all (security team) decide how y'all feel about the different argument order. It is mostly a problem for File-based reading, because there's not different types to the arguments.

**Pull request merged**
